### ISSUE TITLE
Two-Factor Authentication

### ISSUE THREAD

**Issue by hilkbahar** [nuxt-community/auth-module#170]
Hi,

How can I do two-factor authentication ? 
Exp: SMS Auth.

Thank you.

**Comment by angelocapone** [comment]
My I kindly ask to have some details how to implement 2FA with Nuxt?
I'm trying to use the auth module with JWT but I'm unable to make it working...

Thank you
Angelo

**Comment by freddy-daniel** [comment]
Is there any update on this topic ?

**Comment by Phumbie** [comment]
I'm wondering how to implement this as well, if any one can direct me to any link as too

**Comment by freddy-daniel** [comment]
> I'm wondering how to implement this as well, if any one can direct me to any link as too

You can do it with custom auth strategy...

**Comment by Phumbie** [comment]
> > I'm wondering how to implement this as well, if any one can direct me to any link as too
> 
> You can do it with custom auth strategy...

Yes, only custom strategy has been a little confusing for me, the only example showed how to extend loginWith and not how to rewrite your own

**Comment by Myzel394** [comment]
@Phumbie did you find a solution?

**Comment by Adamskoullos** [comment]
Hi there, currently working with nuxt-auth / Laravel Sanctum strategy. Trying to implement 2fa but nuxt-auth will not return the {two_factor: true} object before trying to grab the user and log them in.

Question: Is there a way to prevent nuxt-auth from trying to log the user in and just return the response from the initial login.

I feel, that to get 2fa working a custom auth setup is required....Anyone prove me wrong with nuxt-auth. Thanks

**Comment by sts-ryan-holton** [comment]
Hi, for any one that needs 2FA working, in particular with a third-party library such as Google 2FA, I have this working in my Laravel 8 project, I use Laravel for my back-end, and Nuxt JS for the front-end and spent a long time figuring this out, so thought I'd share my implementation, that's been working really well in a production environment with multiple users for at least 4 months now.

Here's the general flow:

1. First, make sure you're using at least `5.0.0-1620773067.c356fed` of this package.
2. Next, you need to [create a custom scheme](https://auth.nuxtjs.org/guide/scheme#creating-your-own-scheme), create a file called `customScheme.js` inside of a new `schemes` folder, next to all your other folders.
3. You'll be extending the `LocalScheme`, providing you're using the Local scheme that is _(I use the jwt-auth package by Tymondesigns on my backend, this works for that)_, here's my set up:
```javascript
import { LocalScheme } from '~auth/runtime'

export default class CustomScheme extends LocalScheme {

  async login(endpoint, reset = true) {
    if (!this.options.endpoints.login) {
      return
    }

    // Ditch any leftover local tokens before attempting to log in
    if (reset) {
      this.$auth.reset({ resetInterceptor: false })
    }

    // Add client id to payload if defined
    if (this.options.clientId) {
      endpoint.data.client_id = this.options.clientId
    }

    // Add grant type to payload if defined
    if (this.options.grantType) {
      endpoint.data.grant_type = this.options.grantType
    }

    // Add scope to payload if defined
    if (this.options.scope) {
      endpoint.data.scope = this.options.scope
    }

    // Make login request
    const response = await this.$auth.request(
      endpoint,
      this.options.endpoints.login
    )

    // 2FA enabled
    if (
      (response.data.google2fa_enabled != null && response.data.google2fa_enabled == true) ||
      response.data.google2fa_disabled != null
    ) {
      return response
    }

    // Update tokens
    this.updateTokens(response)

    // Initialize request interceptor if not initialized
    if (!this.requestHandler.interceptor) {
      this.initializeRequestInterceptor()
    }

    // Fetch user if `autoFetch` is enabled
    if (this.options.user.autoFetch) {
      await this.fetchUser()
    }

    return response
  }

}
```

Most of you commented about returning some custom logic just before the auth package logs in? Well, instead of returning the token straight away, just return the `response` instead, that way you can conditionally render some kind of 2FA input box on your front-end where upon submission would then validate on the backend and then return a token that way, take note of:

```javascript
// 2FA enabled
if (
  (response.data.google2fa_enabled != null && response.data.google2fa_enabled == true) ||
  response.data.google2fa_disabled != null
) {
  return response
}
```

4. You'll then need to actually use the custom scheme, so add:

```javascript
customStrategy: {
  scheme: '@/schemes/customScheme'
}
```

as a new object inside of `strategies`, and finally update your `this.$auth.loginWith` to be `customScheme`.

-----

 Now, my only issue with this is that this will work fine if you're not planning on using the specific [jwt auth scheme](https://auth.nuxtjs.org/providers/laravel-jwt), I can't seem to get my scheme above to work for that, maybe someone can help me with that?

**Comment by movepixels** [comment]
I am using Laravel 8 as the backend using JWT with Nuxt 2.x front end and I managed to get 2FA working and its quite simple.

This requires no CustomSceme or editing anything in the nuxt/auth plugin.

I used and followed this pretty much except where laravel controllers return a blade i simply return a json response to Nuxt.
```
https://shouts.dev/articles/laravel-two-factor-authentication-with-google-authenticator
```

Now there are 2 fields you need to use to make this work: 
- is 2fa enabled
- has the user been verifed

User logs in as normal but in the response if the user has 2fa enabled send them to a page you need to make which verifies the 2fs code.

```
async userLogin() {
      try {
        const response = await this.$auth.loginWith('laravel', {
          data: this.formData
        })
        this.handleLogin()
      } catch (err) {
        console.log(err)
      }
    },
    handleLogin() {
      if (this.$auth.user.tfa) { // this flags if the user has 2fa enabled, add this in the auth/user request
        this.$router.push({
          name: 'authentication' // some page where user can enter 2fa code
        })
      }
      this.$router.push({ path: this.$auth.user.home }).catch(() => {})
    },
```

On the authentication page you simply need to verify the code and laravel will now if success add a otp => true to the JWT (the magic here)

Nuxt verify method on authentication page
```
methods: {
    verify(url) {
      return this.$validator.validateAll().then(isValid => {
        if (isValid) {
          this.save(url).catch(e => {
            this.handleErrors(e) // server errors
          })
        } else {
          // my custom error functions for client side
          this.scrollToValidationError() 
          this.showError()
        }
      })
    },
    // this gets called auto by my api so you will need to make a function to do this if the verify passes
    handleSaveResponse(response) {
      this.$auth.strategy.token.set(response.refresh_token) // update the auth token sent back from laravel with opt => true
      this.$router.push({ path: this.$auth.user.home }).catch(() => {}) // send the user to their dashboard
    }
  }
```
The laravel verify needs to return an update JWT with otp => true so when you verify the users submitted secret code:
```
$newToken = \JWTAuth::parseToken()->claims(['otp' => true])->refresh();
return response()->json(['valid' => $valid, 'key' => $key, 'refresh_token' => $newToken], 200)->header('Authorization', 'Bearer ' . $newToken);

You can include the Auth header but updating axios does not update the nuxt/Auth token so as of now i return the token in the JSON response

So now you simply build a laravel middleware to check if the user has 2fa enabled and pull the opt value from the JWT to lock the backend

laravel middleware (Can be cleaned up more but it works)
```
public function handle($request, Closure $next)
  {

    /************************************************************************
    * 2FA CHECK
    *************************************************************************
    * ENSURE 2FA PASSED IF ITS ENABLED
    *
    *
    *************************************************************************/

    // DOES THE USER HAVE A 2FA RECORD
    $record = auth()->user()->loginSecurity; // this can be moved to JWT custom claims

    // USER HAS NO RECORD OR HAS A RECORD AND ITS DISABLED -> ALLOW ACCESS
    if(!$record || $record && !$record->google2fa_enable){
      return $next($request);
    }

    // USER HAS A RECORD AND ITS ENABLED
    if($record && $record->google2fa_enable){
      // CHECK JWT THAT ITS BEEN VERIFIED
      $payload = \JWTAuth::parseToken()->getPayload()->get('otp');
      \Log::info('Ensuring user is 2fa verified');
      if($payload){
        return $next($request);
      }
    }

    $response = [
       'message' => 'Unauthorized',
       'text' => 'Illegal Request',
       'code' => '99010'
     ];

     return response()->json($response, 401);

  }
```

And Nuxt middleware:
something as simple as:
```
return app.$axios
      .$get('status', { params:  } }) // some API call for the logged in user or pull it from the auth.Store
      .then(res => {
        const TwoFactorVerified = res.meta.tfa_verified
        const TwoFactorEnabled = res.meta.tfa_enabled

        if (TwoFactorEnabled && !TwoFactorVerified) {
          return redirect('/authentication')
        }
      })
```

So thats how I approached it. No fancy plugins or extensive code.

**Comment by movepixels** [comment]
I am running Laravel 8 backend API and Nuxt 2.x front-end, using JWT  and have implemented the Google2FA with no problems.

No customScheme or anything needed.

Most code can be referenced from here, just skip the Middleware, blade sections, and any response return a json object to Nuxt
[https://shouts.dev/articles/laravel-two-factor-authentication-with-google-authenticator](url)

User logs in and via settings page enables or disables 2FA.
So it works for any user no matter what.
If enabled on login redirects to a "authenticate" page and must enter the 6 digit Google auth code, and middleware locks the API backend and frontend protected with middleware also. So if a user has 2fa enabled they are basically locked on that authenticate page. Attempting any Postman is blocked via the laravel middleware.

Users who do not have 2fa enabled are simple pushed to their dashboard home page.

All you need is 2 keys added to your jwt:
User.php
```
public function getJWTCustomClaims() {
  return [
    '2fa' => [
      'google2fa_enabled' => $this->loginSecurity ? $this->loginSecurity->google2fa_enable : false,
      'google2fa_verified' => false,
    ]
  ];
}
```
Add these same 2fa key / value pairs to your auth/user in laravel response so when a user logs in you can then direct the user to dashboard or authenticate page

```
async userLogin() {
  try {
    const response = await this.$auth.loginWith('laravel', {
      data: this.formData
    })
    this.handleLogin()
  } catch (err) {
    console.log(err)
  }
},
handleLogin() {
  const enabled = this.$auth.user.tfa.google2fa_enabled // you need to pull this from $user->loginSecurity
  if (enabled) {
    this.$router.push({
      name: 'authentication'
    })
  }
  if (!enabled) {
    this.$router.push({ path: this.$auth.user.home })
  }
},
```
If the user verifies the 6 digit code you update the JWT to:
```
'google2fa_enabled' => true,
'google2fa_verified' => true,
```

Return that response with Authorization headers

LoginSecurityController verify method
```
public function verify2fa(Request $request){
    $user = Auth::user();
    $google2fa = (new \PragmaRX\Google2FAQRCode\Google2FA());

    $secret = $request->input('one_time_password');
    $valid = $google2fa->verifyKey($user->loginSecurity->google2fa_secret, $secret);

    if($valid){

      $values = [
        '2fa' => [
          'google2fa_enabled' => true,
          'google2fa_verified' => true
        ]
      ];
      $newToken = \JWTAuth::parseToken()->claims($values)->refresh();
      return response()->json(['valid' => $valid], 200)->header('Authorization', 'Bearer ' . $newToken);
    }else{
      return response()->json(['error' => 'Invalid Google Code'], 422);
    }
}
```
You can intercept the token with axios and update the axios token and the auth token

So you now need to build 2 middleware (laravel , nuxt) that simply check:
1. if !enabled => allow
2.  if enabled and verified => allow
3. if enabled and !verified => not allowed so do something / logout / return to authorization page

Basically need to update the JWT when user enables / disables / verifies the 2fa. Interceptors catch the token change so there is no code changes needed.
Middleware checks the JWT and there you have it 2FA with Google2FA middleware

Hope it helps

**Comment by sts-ryan-holton** [comment]
How can I do Google 2FA with Laravel Sanctum?
